### PR TITLE
gcc: update to version 5.x

### PR DESCRIPTION
**Pull request by noblepepper**
-------------------------------

Maintainer:
Compile tested:
mips Linkit7688, 
OpentWRT Trunk@commit b136759a170bd7348eaebaf0197bf61cf80bf6f3
Lede@commit ea288126dbcf11ecf990b858e1c563b24b198327

arm Broadcom BCM47xx/53xx,
OpentWRT Trunk@commit 9b4650b3b92e6246b986ac9e3d7c2a80d66b805b
Lede@commit ea288126dbcf11ecf990b858e1c563b24b198327

Run tested: 
mips Linkit7688, 
OpentWRT Trunk@commit b136759a170bd7348eaebaf0197bf61cf80bf6f3
Lede@commit ea288126dbcf11ecf990b858e1c563b24b198327

Compiles hello world which runs as expected

Description:
Change Makefile to use the same version as toolchain, uses toolchain patches

### CODE REVIEW

**Comment by noblepepper** [comment]
I should mention that my work is based on the Entware-ng/entware-packages/devel/gcc fork. I made some further changes to produce this pull request but I believe it giving credit where credit is due.

Thanks Entware!

**Comment by hnyman** [comment]
I have doubts about the automatic version bump of the package along the toolchain. It sounds nice, but I am not sure if that would work in reality in the long run. 

Additionally, the toolchain gcc version difference between Openwrt and LEDE would create different gcc package versions, which could possibly lead to confusing error reports & fixes.

**Comment by noblepepper** [comment]
I share your concerns about things changing in the future, a version bump in the OpenWrt toolchain may break things in this package. Hopefully using the toolchain patches will minimize these issues.

@xypron 's [comment](https://github.com/openwrt/packages/issues/2244#issuecomment-174228430) about abi compatibility seems valid to me but I ran into compilation problems in libstdcpp before getting that far (openwrt is on gcc 5.3, lede is on gcc 5.4). I sure this could be handled otherwise but using GCC_VERSION fixed it this time.

To summarize my defense: It works _for now_ and I _plan_ on dealing with they issues as the crop up in the future.

**Comment by noblepepper** [comment]
@hnyman - is this better?

Now uses gcc 5.4.0 

Compiles on mipsel, (ramips mt7688) lede r1725 and openwrt 49941, tested with hello world on target

**Comment by hnyman** [comment]
Looks better. Hardcoding the version gives more stability and matches the general practice to define package versions explicitly.

You should squash (rebase) the four commits into one, so that the PR does not change the same file twice to different directions.

Additionally, please use your real name in the signed-off-by. (in case Noble Pepper is just an username)

(Note:  much of the background discussion is in #2244 , so I reference that here to get it mentioned in this thread).

**Comment by noblepepper** [comment]
@hnyman -
Done.

Yes, Noble Pepper is my real name, strange how it is always available as a user name too ;-)

**Comment by noblepepper** [comment]
@hnyman -The build bots are producing packages for gcc_5.4.0 so I believe [2244](https://github.com/openwrt/packages/issues/2244) [2387](https://github.com/openwrt/packages/issues/2387) & [2581](https://github.com/openwrt/packages/issues/2581) can be closed.
